runtime/compiler/env/VMJ9.cpp

```diff
@@ -7488,120 +7488,110 @@ TR_J9VM::inlineNativeCall(TR::Compilation * comp, TR::TreeTop * callNodeTreeTop,
-                  callerMethod = (J9Method *) callerMethodSymbol->getResolvedMethod()->getPersistentIdentifier();
-                  callerClass = (J9Class*) callerMethodSymbol->getResolvedMethod()->classOfMethod();
-                  }
-
-               skipFrame = transformJlrMethodInvoke(callerMethod, callerClass);
-
-               if (!skipFrame && inlineDepth == stackDepth)
-                  break;
-               if (callerIndex == -1)
-                  break;
-               if (!skipFrame)
-                  inlineDepth++;
-               callerIndex = comp->getInlinedCallSite(callerIndex)._byteCodeInfo.getCallerIndex();
+               TR::ResolvedMethodSymbol *callerMethodSymbol = comp->getJittedMethodSymbol();
+               callerMethod = reinterpret_cast<J9Method *>(callerMethodSymbol->getResolvedMethod()->getPersistentIdentifier());
+               callerClass = reinterpret_cast<J9Class *>(callerMethodSymbol->getResolvedMethod()->classOfMethod());
                }
-
-            if (!skipFrame && (inlineDepth == stackDepth))
+            else
                {
-               bool returnClassLoader;
-               switch (methodID)
-                  {
-                  case TR::sun_reflect_Reflection_getCallerClass:
-                  case TR::java_lang_invoke_MethodHandles_getStackClass:
-                  //case TR::com_ibm_oti_vm_VM_getStackClass:
-                  //case TR::java_lang_Class_getStackClass:
-                     returnClassLoader = false;
-                     break;
-                  default:
-                     returnClassLoader = true;
-                     break;
-                  }
+               callerMethod = reinterpret_cast<J9Method *>(comp->getInlinedCallSite(callerIndex)._methodInfo);
+               callerClass = reinterpret_cast<J9Class *>(comp->getInlinedResolvedMethod(callerIndex)->containingClass());
+               }
 
-               if (returnClassLoader)
-                  {
-                  if (performTransformation(comp, "O^O inlineNativeCall: inline classloader load [%p] for '%s' at bytecode %d\n",
-                        callNode, callNode->getSymbolReference()->getName(comp->getDebug()), callNode->getByteCodeInfo().getByteCodeIndex()))
-                     {
-                     TR::Node::recreate(callNode, TR::aload);
-                     callNode->removeAllChildren();
-                     TR::SymbolReference *callerClassLoaderSymRef = comp->getSymRefTab()->findOrCreateClassLoaderSymbolRef(comp->getMethodSymbol()->getResolvedMethod());
-                     callNode->setSymbolReference(callerClassLoaderSymRef);
-                     }
-                  }
-               else
+            if (callerIndex == -1)
+               break;
+
+            // Skip methods with java.lang.invoke.FrameIteratorSkip annotation
+            if ((J9_ROM_METHOD_FROM_RAM_METHOD(callerMethod)->modifiers & J9AccMethodFrameIteratorSkip) != J9AccMethodFrameIteratorSkip)
```

Comment:
Could you please clarify why this above condition need not be checked if callerIndex == -1 ?

Reply:
We cannot skip past the callerIndex == -1 frame because this is the top level method being compiled. However I think it may be possible that the top-level method is marked with the `FrameIteratorSkip` annotation. In such a case we cannot perform the optimization. I've added an extra check in my [force-push](https://github.com/eclipse-openj9/openj9/compare/026cc75adea21e664a674b7437e87278a0c0a516..f370b1b2c769e5ac2cb8f27f58dfb020404408b7) to fix this.